### ISSUE TITLE
Using Odin Inspector with xNode

### ISSUE THREAD

**Issue by taylank** [Siccity/xNode#168]
Below is what you would have to do if you wanted to use Odin to draw the properties of your nodes:

1) Make NodeEditorBase inherit from OdinEditor
`public abstract class NodeEditorBase<T, A, K> : OdinEditor where A : Attribute, NodeEditorBase<T, A, K>.INodeEditorAttrib where T : NodeEditorBase<T, A, K> where K : ScriptableObject `

2) Add a PropertyTree field to NodeEditorBase
`protected PropertyTree objectTree;`

3) Modify NodeEditor
In OnBodyGUI() method, initialize the property tree right after serializedObject.Update
```csharp
public virtual void OnBodyGUI() {
            
            serializedObject.Update();
            
            // Initialize the property tree. This should be done only once, not every frame.
            if (objectTree == null)
            {
                objectTree = PropertyTree.Create(serializedObject);
            }
```

At this point if you called objectTree.Draw() you would see Odin drawing your fields with their respective attributes. Problem is, xNode will be drawing them too so everything will be rendered twice. Ideally we want xNode to handle the drawing of input and output ports, and Odin to draw everything else.

Further down within OnBodyGUI you will notice a string array named excludes, which xNode uses to skip drawing of some fields. We're going to make use of that. 

4) Move the `string[] excludes = { "m_Script", "graph", "position", "ports" };` line into the class scope. These are things we don't want drawn, neither by xNode, nor by Odin. We also want to tell Odin to not draw properties marked with input and output attributes.

At this point the code should look something like this:
```csharp
        // Do not draw, either with xNode or Odin
        string[] excludes = { "m_Script", "graph", "position", "ports" };
        // let xNode handle these
        string[] drawnbyXNode = {"input", "output"}; 

        public virtual void OnHeaderGUI() {
            GUILayout.Label(target.name, NodeEditorResources.styles.nodeHeader, GUILayout.Height(30));
        }
```

5) Tell xNode what to draw:
Below, in the while loop, we're making xNode skip anything that is not either input or output.

Further down, DrawWithOdin is called to tell Odin which properties to draw. More on that in a bit.

At the very end, we're calling window.Repaint(). This is because some Odin attributes may have foldout animations, or interactive controls that change the size of the node based on property states. Without repaint, the nodes, ports and connections may look off-position as node size changes.

```csharp
 /// <summary> Draws standard field editors for all public fields </summary>
        public virtual void OnBodyGUI() {
            // Unity specifically requires this to save/update any serial object.
            // serializedObject.Update(); must go at the start of an inspector gui, and
            // serializedObject.ApplyModifiedProperties(); goes at the end.
            serializedObject.Update();
            
            // Initialize the property tree. This should be done only once, not every frame.
            if (objectTree == null)
            {
                objectTree = PropertyTree.Create(serializedObject);
            }
            

            // Iterate through serialized properties and draw them like the Inspector (But with ports)
            SerializedProperty iterator = serializedObject.GetIterator();
            //Debug.Log(serializedObject.GetType());
            bool enterChildren = true;
            EditorGUIUtility.labelWidth = 84;
            while (iterator.NextVisible(enterChildren)) {
                enterChildren = false;
                if (drawnbyXNode.Contains(iterator.name)) NodeEditorGUILayout.PropertyField(iterator, true);
            }

            // Iterate through dynamic ports and draw them in the order in which they are serialized
            foreach (XNode.NodePort dynamicPort in target.DynamicPorts) {
                if (NodeEditorGUILayout.IsDynamicPortListPort(dynamicPort)) continue;
                NodeEditorGUILayout.PortField(dynamicPort);
            }
          
            #if ODIN_INSPECTOR
            DrawWithOdin();
            #endif
            
            serializedObject.ApplyModifiedProperties();
            
            // Call repaint so that the graph window elements respond properly to layout changes coming from Odin
            window.Repaint();

        }
```

6) Tell Odin what to draw:
This is the DrawWithOdin method we called above. It iterates over the PropertyTree object, and draws anything that is not covered by xNode and anything in the excludes array.
```csharp
        public void DrawWithOdin()
        {
            InspectorUtilities.BeginDrawPropertyTree(objectTree, true);
            objectTree.EnumerateTree(includeChildren: false).ForEach(p =>
            {
                if (!drawnbyXNode.Contains(p.Name) && !excludes.Contains(p.Name))
                {                    
                    p.Draw();
                }
            });
            InspectorUtilities.EndDrawPropertyTree(objectTree);
        }
```

7) Create a Node class and try out some Odin attributes
Create one of these nodes (or your own) in the node graph and play around with the fields.
```csharp
using System.Collections;
using System.Collections.Generic;
using Sirenix.OdinInspector;
using UnityEngine;
using XNode;

[CreateNodeMenu("FlowNode")]
public class FlowNode : Node
{
    [SerializeField, PreviewField(100, ObjectFieldAlignment.Left), HideLabel] private Sprite sprite;
    [SerializeField, Input] private FlowNode input;
    [SerializeField, Output] private FlowNode output;
    [SerializeField] private bool someBool;
    [SerializeField, TextArea(5, 5), ShowIf("someBool", animate: true)] private string text;
    
    [Sirenix.OdinInspector.Button]
    private void TestFunc()
    {
        
    }

    public override object GetValue(NodePort port)
    {
        return this;
    }
}
```

Hopefully it should look something like this: 
![image](https://user-images.githubusercontent.com/4481766/60362177-1c91ba00-9995-11e9-94bb-49c4c4d27e04.png)

Enjoy!

**Comment by Siccity** [comment]
Added link to this issue in https://github.com/Siccity/xNode/wiki/FAQ#how-do-i-use-odin-inspector-with-xnode

**Comment by TorVestergaard** [comment]
This looks pretty right to me. A few helpful notes from an Odin dev, though:

1) We suggest not repainting every frame, but instead using our extension method to only trigger a window repaint if it has been requested by Odin during drawing - animating layout events will continually request repaints this way until they are done. You would do this by calling window.RepaintIfRequested() or this.RepaintIfRequested(), as it's an extension on both EditorWindow and Editor.

2) If you derive from OdinEditor, you already have a PropertyTree for the editor target in the this.Tree member. However, inheriting from OdinEditor may be a bit superfluous, since in your example code you're not using any of its features regardless.

Cheers! :)

**Comment by taylank** [comment]
Thanks for the feeback, Tor. Yeah I was wondering about Repaint. I'll keep that in mind for the future. And yes, you're right about inheriting from OdinEditor. I was in a hurry to make this work and just threw that in there for good measure. :)

**Comment by leissler** [comment]
> This looks pretty right to me. A few helpful notes from an Odin dev, though:
> 
> 1. We suggest not repainting every frame, but instead using our extension method to only trigger a window repaint if it has been requested by Odin during drawing - animating layout events will continually request repaints this way until they are done. You would do this by calling window.RepaintIfRequested() or this.RepaintIfRequested(), as it's an extension on both EditorWindow and Editor.
> 2. If you derive from OdinEditor, you already have a PropertyTree for the editor target in the this.Tree member. However, inheriting from OdinEditor may be a bit superfluous, since in your example code you're not using any of its features regardless.
> 
> Cheers! :)

I tried this out and using this.RepaintIfRequested() seems to work fine, saving redraw calls. However, trying to use the this.Tree field always returns a null reference.
Out of curiosity I decompiled the OdinEditor class (which is still not documented btw), and it shows that "Tree" is a read-only property with lazy initialization:

`   public PropertyTree Tree
    {
      get
      {
        if (this.tree == null)
        {
          try
          {
            this.tree = PropertyTree.Create(this.serializedObject);
          }
          catch (ArgumentException ex)
          {
          }
        }
        return this.tree;
      }
    }`

It might be that "PropertyTree.Create(this.serializedObject);" fails and returns null.

Any ideas what may cause this?

- Martin

**Comment by taylank** [comment]
What is the object you're creating the property tree of?

**Comment by leissler** [comment]
> What is the object you're creating the property tree of?
When I use the code from taylank with that objectTree variable, everything works
            if (objectTree == null)
            {
                objectTree = PropertyTree.Create(serializedObject);
            }

But when I use "this.Tree" instead of "objectTree" everywhere, then it fails because the Tree property contains null. But I cant initialize this.Tree because its a read-only property and it should be initialized automatically in the getter function of the property, right?

**Comment by taylank** [comment]
Ah I see. Yeah I would expect that to be have been initialized by Odin. Doesn't look like an xNode issue to me. You might wanna ask on Odin's discord channel; they are fairly quick to reply.

**Comment by leissler** [comment]
> Ah I see. Yeah I would expect that to be have been initialized by Odin. Doesn't look like an xNode issue to me. You might wanna ask on Odin's discord channel; they are fairly quick to reply.

Yup, you are right of course. I posted it to the Odin discord in the #programming channel. I had just hoped that TorVestergaard would still read this in here and answer directly.

**Comment by leissler** [comment]
Ok, so after some discussion with Tor, it seems a bad idea to derive from OdinEditor whatsoever.

To make xNode work with Odin, all you would need to do is in **NodeEditorBase**:
```
		#if ODIN_INSPECTOR
		private PropertyTree _objectTree;
		public PropertyTree objectTree
		{
			get
			{
				if (this._objectTree == null)
				{
					try
					{
						this._objectTree = PropertyTree.Create(this.serializedObject);
					}
					catch (ArgumentException ex)
					{
						Debug.Log(ex);
					}
				}
				return this._objectTree;
			}
		}
		#endif
```
And in **NodeEditor**:

```
        /// <summary> Draws standard field editors for all public fields </summary>
        public virtual void OnBodyGUI() {
            // Unity specifically requires this to save/update any serial object.
            // serializedObject.Update(); must go at the start of an inspector gui, and
            // serializedObject.ApplyModifiedProperties(); goes at the end.
            serializedObject.Update();
            string[] excludes = { "m_Script", "graph", "position", "ports" };
            
            #if ODIN_INSPECTOR
            // let xNode handle these
            string[] drawnbyXNode = {"input", "output"}; 
            #endif

            // Iterate through serialized properties and draw them like the Inspector (But with ports)
            SerializedProperty iterator = serializedObject.GetIterator();
            bool enterChildren = true;
            EditorGUIUtility.labelWidth = 84;
            while (iterator.NextVisible(enterChildren)) {
                enterChildren = false;
                #if ODIN_INSPECTOR
                if (drawnbyXNode.Contains(iterator.name)) NodeEditorGUILayout.PropertyField(iterator, true);
                #else
                if (excludes.Contains(iterator.name)) continue;
                NodeEditorGUILayout.PropertyField(iterator, true);
                #endif
            }

            // Iterate through dynamic ports and draw them in the order in which they are serialized
            foreach (XNode.NodePort dynamicPort in target.DynamicPorts) {
                if (NodeEditorGUILayout.IsDynamicPortListPort(dynamicPort)) continue;
                NodeEditorGUILayout.PortField(dynamicPort);
            }

            #if ODIN_INSPECTOR
            InspectorUtilities.BeginDrawPropertyTree(objectTree, true);
            objectTree.EnumerateTree(includeChildren: false).ForEach(p =>
            {
                if (!drawnbyXNode.Contains(p.Name) && !excludes.Contains(p.Name))
                {                    
                    p.Draw();
                }
            });
            InspectorUtilities.EndDrawPropertyTree(objectTree);
            #endif
            
            serializedObject.ApplyModifiedProperties();
            
            // Call repaint so that the graph window elements respond properly to layout changes coming from Odin    
            #if ODIN_INSPECTOR
            if (GUIHelper.RepaintRequested)
            {
                GUIHelper.ClearRepaintRequest();
            }
            #else
            window.Repaint();
            #endif
        }
```

**Comment by leissler** [comment]
Could be a pull request?

**Comment by Siccity** [comment]
Does Odin add the ODIN_INSPECTOR preprocessor symbol automatically or will i have to instruct people to add it?

**Comment by leissler** [comment]
Odin adds it when you install it.

**Comment by Siccity** [comment]
I guess we could add this then, if this solves the Odin-related questions

**Comment by leissler** [comment]
Although it looks correct, I have only tested it with Odin installed. You should test it with plain Unity too of course. I'm not worried though.

**Comment by Siccity** [comment]
Yeah i'm trying it out on my machine right now. If it works i'll ask some people on the discord to test it, then merge it

**Comment by Siccity** [comment]
One question; it looks like you added Odins draw calls _after_ the dynamic port draw loop. Is there any reason for that? Otherwise i'd switch them around for consistency

**Comment by Siccity** [comment]
https://github.com/Siccity/xNode/tree/experimental/odin-inspector-support
Created a test branch

**Comment by leissler** [comment]
No specific reason for that. The original modifications from taylank had a method DrawWithOdin() called at that point in the code. I just inlined it so you can keep the "excludes" and "drawnbyXNode" arrays local.

**Comment by Siccity** [comment]
I tested it and the ports stopped drawing.
Looked through the code and found these lines:
```cs
string[] drawnbyXNode = { "input", "output" };
```

```cs
if (drawnbyXNode.Contains(iterator.name)) NodeEditorGUILayout.PropertyField(iterator, true);
```

Am i correct that this code assumes the names of the ports being "input" and "output"? Because that's not viable at all.

I swapped the first line with 
```cs
string[] drawnbyXNode = target.Ports.Select(x => x.fieldName).ToArray();
```
and it made some improvements, but i'm still not sure how all this works in the long run

**Comment by TorVestergaard** [comment]
Looks like you guys are missing a window.Repaint(); call in the if (GUIHelper.RepaintRequested) clause as well - currently it's not triggering any repaints, it looks like.

**Comment by Siccity** [comment]
Added my change to drawnbyXNode as well as your window.Repaint. It seems pretty solid now. The only caveat is that none of the attributes work on port fields

**Comment by Siccity** [comment]
> We would be interested to use this in our project since we're using odin pretty much extensively. I've done the change by hand but if there is a branch somewhere with the "official" fix that need to be tested I can help.

There is actually. I'd appreciate some testing on it :)
https://github.com/Siccity/xNode/tree/experimental/odin-inspector-support

**Comment by DaveSirois** [comment]
Hi,

I would be interested to use this. This link doesn't seem to work:
https://github.com/Siccity/xNode/tree/experimental/odin-inspector-support

And I don't find experimental branch for it too.

Thanks :)

**Comment by Siccity** [comment]
> Hi,
> 
> I would be interested to use this. This link doesn't seem to work:
> https://github.com/Siccity/xNode/tree/experimental/odin-inspector-support
> 
> And I don't find experimental branch for it too.
> 
> Thanks :)

Support for Odin Inspector has been merged into the master branch.
